vault_service/views/utils.py

```diff
@@ -85,7 +91,483 @@ def check_request(request):
+            phrase = phrase + ' ' + key_list[i]
+            i += 1
+        first = phrase + ' ' + key_list[i]
+
+    if first != keywords:
+        first = first + ', etc.'
+
+    return first
+
+
+def parse_classic_keywords(query):
+    """
+    Wrapper function to parse the Classic keyword string and return a BBB-style keyword string
+
+    :param query: string; Classic-style keyword query
+    :return: string; BBB-style keyword query
+    """
+    clean_query = query.strip()
+    tree = _parse_classic_keywords_to_tree(clean_query)
+
+    v = TreeVisitor()
+    new_query = v.visit(tree).output
+
+    return new_query
+
+
+def _parse_classic_keywords_to_tree(data):
+    """
+    Given a string of keywords from Classic, parse the query tree
+
+    :param data: string of Classic keywords
+    :return: parsed tree
+    """
+
+    grammar = Lark(r"""
+
+    
+    start: clause+ (operator clause)*
+    
+    clause: ("(" clause (operator? clause)* ")") 
+        | query+
+
+    query: qterm
+    
+    qterm: anyterm -> qterm | phrase | PREPEND -> prepend
+    
+    PREPEND.2: /=/ | /\+/ | /\-/
+        
+    phrase: DOUBLE_QUOTED_STRING | SINGLE_QUOTED_STRING
+
+    DOUBLE_QUOTED_STRING.3  : /"[^"]*"/ | /\+"[^"]*"/ | /\-"[^"]*"/
+    SINGLE_QUOTED_STRING.3  : /'[^']*'/ | /\+'[^']*'/ | /\-'[^']*'/
+    
+    anyterm: /[^)^\] \(^\n^\r]+/
+    
+    operator: OPERATOR | NEWLINE
+
+    OPERATOR.2: "and" | "AND" | "or" | "OR" | "not" | "NOT" | "AND NOT" | "and not" | /,/ 
+
+    %import common.LETTER
+    %import common.ESCAPED_STRING
+    %import common.FLOAT
+    %import common.DIGIT
+    %import common.WS_INLINE
+    %import common.NEWLINE
+    
+    %ignore WS_INLINE
+    
+    """, parser="lalr")
+
+    tree = grammar.parse(data)
+
+    return tree
+
+
+class TreeVisitor(Visitor):
+    """
+    Visitor class to transform the parsed tree into a BBB-style query.
+    The final constructed query is stored in v.visit(tree).output
+    """
+    def start(self, node):
+        out = []
+        for x in node.children:
+            if hasattr(x, 'output'):
+                out.append(getattr(x, 'output'))
+            else:
+                pass
+        tmp = ' '.join(out)
+
+        node.output = tmp
+
+    def clause(self, node):
+        out = []
+        ops = ['and', 'AND', 'or', 'OR', 'not', 'NOT', 'and not', 'AND NOT']
+        prepend = ['=', '+', '-']
+        for x in node.children:
+            if hasattr(x, 'output'):
+                out.append(getattr(x, 'output'))
+        output = []
+        i = 0
+        for o in out:
+            if i == 0:
+                output.append(o)
+            else:
+                if output[i-1] in ops:
+                    output.append(o)
+                elif output[i-1] in prepend:
+                    if i < 2:
+                        output[i-1] += o
+                    else:
+                        output[i-1] = 'OR ' + output[i-1] + o
+                    output.append('')
```

Comment:
line 537 surprised me little bit, it might be saner to format the `output[i-1]` entry instead of inserting an empty placeholder (for space)